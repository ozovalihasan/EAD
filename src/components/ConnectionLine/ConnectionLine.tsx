import { ConnectionLineComponentProps, getBezierPath, Node } from "react-flow-renderer";
import useStore from "@/zustandStore/store";

export const ConnectionLine = ({
  sourceX,
  sourceY,
  sourcePosition,
  targetX,
  targetY,
  targetPosition,
}: Omit<ConnectionLineComponentProps, 'connectionLineType'>) => {

  const associationType = useStore(state => state.associationType)

  let throughNode: null | Node = null
  let throughEdgePath: null | string = null

  if (associationType === 'through'){
    const selectedNodeIdForThrough = useStore(state => state.selectedNodeIdForThrough)

    throughNode = useStore(store => store.nodes.find( node => node.id === selectedNodeIdForThrough)) as Node
    
    if (throughNode){
      throughEdgePath = getBezierPath({
        sourceX,
        sourceY,
        sourcePosition,
        targetX: throughNode.position.x + (throughNode.width! / 2),
        targetY: throughNode.position.y + (throughNode.height! / 2),
        targetPosition,
        curvature: 0.5
      })
    }
  }

  let sourceUpdatedX = null
  let sourceUpdatedY = null

  if (throughNode) {
    sourceUpdatedX = throughNode.position.x + (throughNode.width! / 2);
    sourceUpdatedY = throughNode.position.y + (throughNode.height! / 2);
  } else{
  sourceUpdatedX = sourceX
    sourceUpdatedY = sourceY
  }

  const edgePath = getBezierPath({
    sourceX: sourceUpdatedX,
    sourceY: sourceUpdatedY,
    sourcePosition,
    targetX,
    targetY,
    targetPosition,
    curvature: 0.5
  })

  
  
  return (
    <>
      <g>
        {
          throughEdgePath &&
          <path
            fill="none"
            stroke="#222"
            strokeWidth={1.5}
            className="custom-animation"
            d={throughEdgePath}
          />
        }
      
        <path
          fill="none"
          stroke="#222"
          strokeWidth={1.5}
          className={associationType === 'through' ? "custom-animation" : ""}
          d={edgePath}
        />
      
        <circle cx={targetX} cy={targetY} fill="#fff" r={3} stroke="#222" strokeWidth={1.5} />
      </g>
    </>
  );
};
