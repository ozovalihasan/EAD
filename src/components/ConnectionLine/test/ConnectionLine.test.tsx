import { ConnectionLine } from '../ConnectionLine';
import { render, screen } from "@testing-library/react";
import useStore from '@/zustandStore/store';
import { Position } from 'react-flow-renderer';



let renderReadyComponent: JSX.Element;

beforeEach(() => {
  
  renderReadyComponent = (
    <svg>
      <ConnectionLine
        sourceX={10}
        sourceY={20}
        sourcePosition={Position.Bottom}
        targetX={30}
        targetY={40}
        targetPosition={Position.Top}
      />
    </svg>
  );
});

describe('<ConnectionLine />', () => {
  describe('if association type is "through"', () => {
    
    it('renders correctly', () => {
      useStore.setState({ 
        associationType: 'through',
        nodes: [
          {
            "id": "1",
            "type": "mockType",
            "position": {
                "x": 465,
                "y": 150
            },
            "data": {
                "tableId": "1",
                "name": "mockName"
            },
            "width": 138,
            "height": 45
          },
        ],
        selectedNodeIdForThrough: "1"
      })
      
      const renderedContainer = render(renderReadyComponent );
      expect(renderedContainer).toMatchSnapshot();
    });
    
  })

  describe('if association type is"not "through"', () => {
    it('renders correctly', () => {
      useStore.setState({ 
        associationType: 'has_many'
      })
      
      const renderedContainer = render(renderReadyComponent );
      expect(renderedContainer).toMatchSnapshot();
    });
    
  })

});