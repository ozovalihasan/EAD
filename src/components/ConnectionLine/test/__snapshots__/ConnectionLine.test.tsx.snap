// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConnectionLine /> if association type is "through" renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <svg>
        <g>
          <path
            class="animated"
            d="M10,20 C10,96.25 534,96.25 534,172.5"
            fill="none"
            stroke="#222"
            stroke-width="1.5"
          />
          <path
            class="animated"
            d="M534,172.5 C534,316.38580541526676 30,-103.88580541526673 30,40"
            fill="none"
            stroke="#222"
            stroke-width="1.5"
          />
          <circle
            cx="30"
            cy="40"
            fill="#fff"
            r="3"
            stroke="#222"
            stroke-width="1.5"
          />
        </g>
      </svg>
    </div>
  </body>,
  "container": <div>
    <svg>
      <g>
        <path
          class="animated"
          d="M10,20 C10,96.25 534,96.25 534,172.5"
          fill="none"
          stroke="#222"
          stroke-width="1.5"
        />
        <path
          class="animated"
          d="M534,172.5 C534,316.38580541526676 30,-103.88580541526673 30,40"
          fill="none"
          stroke="#222"
          stroke-width="1.5"
        />
        <circle
          cx="30"
          cy="40"
          fill="#fff"
          r="3"
          stroke="#222"
          stroke-width="1.5"
        />
      </g>
    </svg>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<ConnectionLine /> if association type is"not "through" renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <svg>
        <g>
          <path
            class=""
            d="M10,20 C10,30 30,30 30,40"
            fill="none"
            stroke="#222"
            stroke-width="1.5"
          />
          <circle
            cx="30"
            cy="40"
            fill="#fff"
            r="3"
            stroke="#222"
            stroke-width="1.5"
          />
        </g>
      </svg>
    </div>
  </body>,
  "container": <div>
    <svg>
      <g>
        <path
          class=""
          d="M10,20 C10,30 30,30 30,40"
          fill="none"
          stroke="#222"
          stroke-width="1.5"
        />
        <circle
          cx="30"
          cy="40"
          fill="#fff"
          r="3"
          stroke="#222"
          stroke-width="1.5"
        />
      </g>
    </svg>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
